File: app/src/main/java/com/zionhuang/music/extractor/JSInterpreter.java
package com.zionhuang.music.extractor;

import androidx.core.util.Pair;

import com.google.gson.JsonParser;
import com.google.gson.JsonSyntaxException;
import com.zionhuang.music.utils.Utils;

import org.apache.commons.lang3.StringUtils;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.StringJoiner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static java.lang.Math.min;

public class JSInterpreter {
    public static class InterpretException extends java.lang.Exception {
        InterpretException(String msg) {
            super(msg);
        }
    }

    private interface OperatorFunction {
        JSON apply(JSON a, JSON b);
    }

    private static LinkedHashMap<String, OperatorFunction> OPERATORS = new LinkedHashMap<>();
    private static LinkedHashMap<String, OperatorFunction> ASSIGN_OPERATORS = new LinkedHashMap<>();

    static {
        OPERATORS.put("|", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() | b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put("^", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() ^ b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put("&", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() & b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put(">>", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() >> b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put("<<", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() << b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put("-", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() - b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put("+", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() + b.getAsInt());
            }
            if (a.isString() && b.isString()) {
                return JSON.fromString(a.getAsString() + b.getAsString());
            }
            return JSON.NULL;
        });
        OPERATORS.put("%", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() % b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put("/", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() / b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.put("*", (a, b) -> {
            if (a.isNumber() && b.isNumber()) {
                return JSON.fromNumber(a.getAsInt() * b.getAsInt());
            }
            return JSON.NULL;
        });
        OPERATORS.forEach((key, value) -> ASSIGN_OPERATORS.put(key + "=", value));
        ASSIGN_OPERATORS.put("=", (a, b) -> b);
    }

    private String code;
    private JSON objects;
    private HashMap<String, JSFunction> functions;

    public JSInterpreter(String code) {
        this.code = code;
        this.objects = JSON.createObject();
        this.functions = new HashMap<>();
    }

    private Pair<JSON, Boolean> interpretStatement(String stmt, JSON localVars) throws InterpretException {
        return interpretStatement(stmt, localVars, 100);
    }

    private Pair<JSON, Boolean> interpretStatement(String stmt, JSON localVars, int allowRecursion) throws InterpretException {
        if (allowRecursion < 0) {
            throw new InterpretException("Recursion limit reached");
        }
        boolean abort = false;
        String expr;
        stmt = stmt.trim();
        Matcher m;
        m = Pattern.compile("^var\\s").matcher(stmt);
        if (m.find()) {
            expr = stmt.substring(m.end(0));
        } else {
            m = Pattern.compile("^return(?:\\s+|$)").matcher(stmt);
            if (m.find()) {
                expr = stmt.substring(m.end(0));
                abort = true;
            } else {
                expr = stmt;
            }
        }
        return new Pair<>(interpretExpression(expr, localVars, allowRecursion), abort);
    }

    private JSON interpretExpression(String expr, JSON localVars, int allowRecursion) throws InterpretException {
        expr = expr.trim();
        if (expr.isEmpty()) {
            return JSON.NULL;
        }

        Matcher m;
        if (expr.startsWith("(")) {
            int parenCount = 0;
            boolean found = false;
            m = Pattern.compile("[()]").matcher(expr);
            while (m.find()) {
                found = true;
                if (m.group().equals("(")) {
                    parenCount += 1;
                } else {
                    parenCount -= 1;
                    if (parenCount == 0) {
                        String subExpr = expr.substring(1, m.start());
                        JSON subResult = interpretExpression(subExpr, localVars, allowRecursion);
                        String remainingExpr = expr.substring(m.end()).trim();
                        if (remainingExpr.isEmpty()) {
                            return subResult;
                        } else {
                            expr = subResult.toString() + remainingExpr;
                        }
                        break;
                    }
                }
            }
            if (!found) {
                throw new InterpretException("Premature end of parenthesis at " + expr);
            }
        }

        for (Map.Entry<String, OperatorFunction> entry : ASSIGN_OPERATORS.entrySet()) {
            m = Pattern.compile("(?x)(?<out>[a-zA-Z_$][a-zA-Z_$0-9]*)(?:\\[(?<index>[^\\]]+?)\\])?\\s*%s(?<expr>.*)$".replace("%s", Pattern.quote(entry.getKey()))).matcher(expr);
            if (!m.matches()) {
                continue;
            }
            JSON rightVal = interpretExpression(m.group("expr"), localVars, allowRecursion - 1);
            if (m.group("index") != null) {
                int idx = interpretExpression(m.group("index"), localVars, allowRecursion).getAsInt();
                JSON lVar = localVars.get(m.group("out"));
                if (!lVar.isArray()) {
                    throw new InterpretException("Unsupported operator [] in non-array object at " + expr);
                }
                JSON val = entry.getValue().apply(lVar.get(idx), rightVal);
                lVar.set(idx, val);
                return val;
            } else {
                JSON cur = localVars.get(m.group("out"));
                JSON val = entry.getValue().apply(cur, rightVal);
                localVars.add(Objects.requireNonNull(m.group("out")), val);
                return val;
            }
        }

        if (StringUtils.isNumeric(expr)) {
            return JSON.fromNumber(Integer.parseInt(expr));
        }

        m = Pattern.compile("(?!if|return|true|false)(?<name>[a-zA-Z_$][a-zA-Z_$0-9]*)$").matcher(expr);
        if (m.matches()) {
            return localVars.get(m.group("name"));
        }

        try {
            // check if is string first, because method call will be parsed into string
            for (char quote : new char[]{'\"', '\''}) {
                if (expr.charAt(0) == quote && expr.charAt(expr.length() - 1) == quote) {
                    expr = expr.substring(1, expr.length() - 1);
                    return JSON.fromString(expr);
                }
            }
            JSON res = new JSON(JsonParser.parseString(expr));
            if (!res.isString()) {
                return res;
            }
        } catch (JsonSyntaxException ignored) {
        }

        m = Pattern.compile("(?<in>[a-zA-Z_$][a-zA-Z_$0-9]*)\\[(?<idx>.+)\\]$").matcher(expr);
        if (m.matches()) {
            JSON val = localVars.get(m.group("in"));
            JSON idx = interpretExpression(m.group("idx"), localVars, allowRecursion - 1);
            if (val.isObject()) {
                return val.get(idx.getAsString());
            } else if (val.isArray()) {
                return val.get(idx.getAsInt());
            } else if (val.isString()) {
                return JSON.fromChar(val.getAsString().charAt(idx.getAsInt()));
            } else {
                throw new InterpretException("Unsupported operation [] for expression " + expr);
            }
        }

        m = Pattern.compile("(?<var>[a-zA-Z_$][a-zA-Z_$0-9]*)(?:\\.(?<member>[^(]+)|\\[(?<member2>[^]]+)\\])\\s*(?:\\(+(?<args>[^()]*)\\))?$").matcher(expr);
        if (m.matches()) {
            String variable = m.group("var");
            String member = Utils.removeQuotes(m.group("member") != null ? m.group("member") : m.group("member2"));
            String argStr = m.group("args");

            Objects.requireNonNull(variable);
            JSON obj;
            if (localVars.has(variable)) {
                obj = localVars.get(variable);
            } else {
                if (!objects.has(variable)) {
                    objects.add(variable, extractObject(variable));
                }
                obj = objects.get(variable);
            }

            if (argStr == null) {
                if (Objects.equals(member, "length")) {
                    if (obj.isArray()) {
                        return JSON.fromNumber(obj.size());
                    } else if (obj.isString()) {
                        return JSON.fromNumber(obj.getAsString().length());
                    } else {
                        return JSON.NULL;
                    }
                }
                if (obj.isArray()) {
                    return obj.get(Integer.parseInt(Objects.requireNonNull(member)));
                } else if (obj.isObject()) {
                    return obj.get(member);
                } else if (obj.isString()) {
                    return JSON.fromChar(obj.getAsString().charAt(Integer.parseInt(Objects.requireNonNull(member))));
                } else {
                    return JSON.NULL;
                }
            }

            if (!expr.endsWith(")")) {
                throw new InterpretException("Expression missing ) at " + expr);
            }

            // Function call
            JSON argValues = JSON.createArray();
            if (argStr.length() > 0) {
                for (String v : argStr.split(",")) {
                    argValues.add(interpretExpression(v, localVars, allowRecursion));
                }
            }

            if (Objects.equals(member, "split")) {
                if (!obj.isString()) {
                    throw new InterpretException("Calling split function on a non-string object");
                }
                if (argValues.getString(0).length() != 0) {
                    throw new InterpretException("Parameter error at " + expr);
                }
                JSON res = JSON.createArray();
                for (char c : obj.getAsString().toCharArray()) {
                    res.add(c);
                }
                return res;
            }
            if (Objects.equals(member, "join")) {
                if (!obj.isArray()) {
                    throw new InterpretException("Calling join function on non-array type at " + expr);
                }
                if (argValues.size() != 1) {
                    throw new InterpretException("Parameter error at " + expr);
                }
                if (!argValues.get(0).isString()) {
                    throw new InterpretException("Error argument type at " + expr);
                }
                StringJoiner joiner = new StringJoiner(argValues.get(0).getAsString());
                obj.forEach(ele -> joiner.add(Utils.removeQuotes(ele.toString())));
                return JSON.fromString(joiner.toString());
            }
            if (Objects.equals(member, "reverse")) {
                if (!obj.isArray()) {
                    throw new InterpretException("Calling reverse function on non-array type at " + expr);
                }
                if (argValues.size() != 0) {
                    throw new InterpretException("Parameter error at " + expr);
                }
                int size = obj.size();
                // swap
                for (int i = 0; i < (size >> 1); i++) {
                    JSON temp = obj.get(i);
                    obj.set(i, obj.get(size - i - 1));
                    obj.set(size - i - 1, temp);
                }
                return obj;
            }
            if (Objects.equals(member, "slice")) {
                if (argValues.size() != 1) {
                    throw new InterpretException("Parameter error at " + expr);
                }
                return JSON.fromString(obj.getAsString().substring(argValues.getInt(0)));
            }
            if (Objects.equals(member, "splice")) {
                if (!obj.isArray()) {
                    throw new InterpretException("Can't call a function splice which is not a JsonArray type at " + expr);
                }
                int index = argValues.get(0).getAsInt();
                int howMany = argValues.get(1).getAsInt();
                JSON res = JSON.createArray();
                for (int i = index; i < min(index + howMany, obj.size()); i++) {
                    res.add(obj.remove(index));
                }
                return res;
            }

            // custom function
            if (obj.isObject()) {
                JSON fn = obj.get(member);
                if (!fn.isFunction()) {
                    throw new InterpretException("Calling function on a non-function object at " + expr);
                }
                return fn.getAsFunction().apply(argValues);
            }
        }

        for (Map.Entry<String, OperatorFunction> entry : OPERATORS.entrySet()) {
            m = Pattern.compile("(?<x>.+?)" + Pattern.quote(entry.getKey()) + "(?<y>.+)").matcher(expr);
            if (!m.matches()) {
                continue;
            }
            Pair<JSON, Boolean> pairX = interpretStatement(m.group("x"), localVars, allowRecursion - 1);
            if (Objects.requireNonNull(pairX.second)) {
                throw new InterpretException("Premature left-side return of " + entry.getKey() + " at " + expr);
            }
            Pair<JSON, Boolean> pairY = interpretStatement(m.group("y"), localVars, allowRecursion - 1);
            if (Objects.requireNonNull(pairY.second)) {
                throw new InterpretException("Premature right-side return of " + entry.getKey() + " at " + expr);
            }
            return entry.getValue().apply(pairX.first, pairY.first);
        }

        m = Pattern.compile("^(?<func>[a-zA-Z_$][a-zA-Z_$0-9]*)\\((?<args>[a-zA-Z0-9_$,]*)\\)$").matcher(expr);
        if (m.matches()) {
            String funcName = m.group("func");
            String args = m.group("args");
            Objects.requireNonNull(args);

            JSON argValues = JSON.createArray();
            if (args.length() > 0) {
                for (String v : args.split(",")) {
                    argValues.add(StringUtils.isNumeric(v) ? JSON.fromNumber(Integer.parseInt(v)) : localVars.get(v));
                }
            }
            if (!functions.containsKey(funcName)) {
                functions.put(funcName, extractFunction(funcName));
            }
            return Objects.requireNonNull(functions.get(funcName)).apply(argValues);
        }
        throw new InterpretException("Unsupported JS expression " + expr);
    }

    private JSON extractObject(String objName) throws InterpretException {
        JSON obj = JSON.createObject();
        Matcher objM = Pattern.compile("(?x)(?<!this\\.)" + objName + "\\s*=\\s*\\{\\s*(?<fields>((?:[a-zA-Z$0-9]+|\"[a-zA-Z$0-9]+\"|'[a-zA-Z$0-9]+')\\s*:\\s*function\\s*\\(.*?\\)\\s*\\{.*?\\}(?:,\\s*)?)*)\\}\\s*;").matcher(code);
        if (!objM.find()) {
            throw new InterpretException("Failed to extract object " + objName);
        }
        String fields = objM.group("fields");
        Objects.requireNonNull(fields);

        Matcher fieldsM = Pattern.compile("(?x)(?<key>(?:[a-zA-Z$0-9]+|\"[a-zA-Z$0-9]+\"|'[a-zA-Z$0-9]+'))\\s*:\\s*function\\s*\\((?<args>[a-z,]+)\\)\\{(?<code>[^}]+)\\}").matcher(fields);
        while (fieldsM.find()) {
            String[] argNames = Objects.requireNonNull(fieldsM.group("args")).split(",");
            obj.add(Objects.requireNonNull(Utils.removeQuotes(fieldsM.group("key"))), JSON.fromFunction(buildFunction(argNames, fieldsM.group("code"))));
        }
        return obj;
    }

    public JSFunction extractFunction(String funcName) throws InterpretException {
        String escapedFnName = Pattern.quote(funcName);
        Matcher m = Pattern.compile("(?x)(?:function\\s+" + escapedFnName + "|[{;,]\\s*" + escapedFnName + "\\s*=\\s*function|var\\s+" + escapedFnName + "\\s*=\\s*function)\\s*\\((?<args>[^)]*)\\)\\s*\\{(?<code>[^}]+)\\}").matcher(code);
        if (!m.find()) {
            throw new InterpretException("Couldn't find JS function " + funcName);
        }
        String[] argNames = Objects.requireNonNull(m.group("args")).split(",");
        return buildFunction(argNames, m.group("code"));
    }

    public JSON callFunction(String funcName, Object... args) throws InterpretException {
        return callFunction(funcName, JSON.toArray(args));
    }

    public JSON callFunction(String funcName, JSON args) throws InterpretException {
        return extractFunction(funcName).apply(args);
    }

    private JSFunction buildFunction(final String[] argNames, final String code) {
        return new JSFunction() {
            @Override
            public JSON apply(JSON args) throws InterpretException {
                JSON localVars = JSON.createObject();
                for (int i = 0; i < args.size() && i < argNames.length; i++) {
                    localVars.add(argNames[i], args.get(i));
                }
                Pair<JSON, Boolean> pair;
                JSON res = JSON.NULL;
                for (String stmt : code.split(";")) {
                    pair = interpretStatement(stmt, localVars);
                    res = pair.first;
                    if (Objects.requireNonNull(pair.second)) {
                        break;
                    }
                }
                return res;
            }
        };
    }
}

File: app/src/main/java/com/zionhuang/music/youtube/utils/JsInterpreter.kt
@file:Suppress("RegExpRedundantEscape")

package com.zionhuang.music.youtube.utils

import com.google.gson.*
import com.zionhuang.music.extensions.*
import okhttp3.internal.toImmutableMap
import org.intellij.lang.annotations.Language

typealias OperatorFunction = (JsonElement, JsonElement) -> JsonElement

class JsInterpreter(private val code: String) {
    private val objects = JsonObject()
    private val functions = HashMap<String, JsFunction>()

    @Throws(InterpretException::class)
    private fun interpretStatement(stmt: String, localVars: JsonObject, allowRecursion: Int = 100): Pair<JsonElement, Boolean> {
        if (allowRecursion < 0) throw InterpretException("Recursion limit exceed")
        var abort = false
        val expr: String = with(stmt.trim()) {
            find(VAR_RE)?.let { substring(it.range.last + 1) }
                    ?: find(RETURN_RE)?.let { abort = true;substring(it.range.last + 1) }
                    ?: this
        }
        return Pair(interpretExpression(expr, localVars, allowRecursion), abort)
    }

    @Throws(InterpretException::class)
    private fun interpretExpression(originExpr: String, localVars: JsonObject, allowRecursion: Int): JsonElement {
        var expr = originExpr.trim()
        if (expr.isEmpty()) JsonNull.INSTANCE
        if (expr.startsWith("(")) {
            var parenCount = 0
            var found = false
            for (match in expr.findAll("""[()]""")) {
                found = true
                if (match.value == "(") {
                    parenCount++
                } else {
                    parenCount--
                    if (parenCount == 0) {
                        val subExpr = expr[1, match.range.first]
                        val subResult = interpretExpression(subExpr, localVars, allowRecursion)
                        val remainingExpr = expr.substring(match.range.last + 1).trim()
                        if (remainingExpr.isEmpty()) {
                            return subResult
                        } else {
                            expr = subResult.toString() + remainingExpr
                        }
                        break
                    }
                }
            }
            if (!found) {
                throw InterpretException("Premature end of parenthesis at $expr")
            }
        }

        for ((op, opFn) in ASSIGN_OPERATORS) {
            val match = expr.matchEntire(ASSIGN_OPERATOR_RE % op.escape()) ?: continue
            val group = match.groups
            val rightVal = interpretExpression(group["expr"]!!.value, localVars, allowRecursion - 1)
            return if (group["index"] != null) {
                val idx = interpretExpression(group["index"]!!.value, localVars, allowRecursion).asIntOrNull
                        ?: throw InterpretException("Index is not a integer: $expr")
                val lVar = localVars[group["out"]!!.value].asJsonArrayOrNull
                        ?: throw InterpretException("Expected an array object at the left-hand value: $expr")
                opFn(lVar[idx], rightVal).also {
                    lVar[idx] = it
                }
            } else {
                val lVarName = group["out"]!!.value
                val lVar = localVars[lVarName] ?: JsonNull.INSTANCE
                opFn(lVar, rightVal).also {
                    localVars[lVarName] = it
                }
            }
        }

        expr.toIntOrNull()?.let { return it.toJsonPrimitive() }

        expr.matchEntire(VARIABLE_RE, "name")?.let {
            return localVars[it] ?: throw InterpretException("variable $it not found")
        }

        if (expr.length > 1) {
            for (quote in arrayOf('\"', '\'')) {
                if (expr.first() == quote && expr.last() == quote) {
                    return expr[1, expr.length - 1].toJsonPrimitive()
                }
            }
        }

        try {
            // whenever the expression is not a string type, parseJsonString always parse it into a string
            val parsed = expr.parseJsonString()
            if (!parsed.isString()) return parsed
        } catch (e: JsonSyntaxException) {
        }

        expr.matchEntire(VARIABLE_INDEX_RE)?.let { match ->
            val lVar = localVars[match.groupValue("in")!!]
            val idx = interpretExpression(match.groupValue("idx")!!, localVars, allowRecursion - 1)
            return when {
                idx.isString() -> lVar[idx.asString]
                idx.isNumber() -> lVar[idx.asInt]
                else -> throw InterpretException("index is neither string nor int")
            } ?: throw InterpretException("property $idx doesn't exist")
        }

        expr.matchEntire(MEMBER_RE)?.let { match ->
            val varName = match.groupValue("var")!!
            val obj = localVars[varName] ?: extractObject(varName).also {
                objects[varName] = it
            }
            val member = (match.groupValue("member")
                    ?: match.groupValue("member2")!!).removeQuotes()
            val argStr = match.groupValue("args")

            if (argStr == null) {
                if (member == "length") {
                    return when {
                        obj is JsonArray -> obj.size()
                        obj is JsonPrimitive && obj.isString -> obj.asString.length
                        else -> throw InterpretException("""$varName has no member "length"""")
                    }.toJsonPrimitive()
                }
                return obj.asJsonObjectOrNull[member]
                        ?: throw InterpretException("""$varName doesn't exist or doesn't have the member "$member"""")
            }

            // Function call
            val argValues = argStr.split(",").map { interpretExpression(it, localVars, allowRecursion) }.toJsonArray()
            return when (member) {
                "split" -> {
                    // assert argValues == [""]
                    obj.asString.splitToJsonArray()
                }
                "join" -> {
                    // assert argValues.length == 1
                    obj.asJsonArray.joinToString(argValues[0].asString) {
                        it.asString.removeQuotes()
                    }.toJsonPrimitive()
                }
                "reverse" -> {
                    obj.asJsonArray.selfReverse()
                }
                "slice" -> {
                    obj.asString.substring(argValues[0].asInt).toJsonPrimitive()
                }
                "splice" -> {
                    val (index, howMany) = argValues.map { it.asInt }
                    if (obj !is JsonArray) throw InterpretException("$varName is not an object")
                    JsonArray().apply {
                        for (i in index until minOf(index + howMany, obj.size())) {
                            add(obj.remove(index))
                        }
                    }
                }
                else -> {
                    (obj[member]!! as JsFunction)(argValues)
                }
            }
        }

        for ((op, opFn) in OPERATORS) {
            val match = expr.matchEntire("""(?<x>.+?)${op.escape()}(?<y>.+)""") ?: continue
            val (x, lAbort) = interpretStatement(match.groupValue("x")!!, localVars, allowRecursion - 1)
            if (lAbort) throw InterpretException("Premature left-side return of $op at $expr")
            val (y, rAbort) = interpretStatement(match.groupValue("y")!!, localVars, allowRecursion - 1)
            if (rAbort) throw InterpretException("Premature right-side return of $op at $expr")
            return opFn(x, y)
        }

        expr.matchEntire(FN_CALL_RE)?.let { match ->
            val fnName = match.groupValue("func")!!
            val argStr = match.groupValue("args")!!
            val argValues = argStr.split(",").mapToJsonArray {
                it.toIntOrNull()?.toJsonPrimitive() ?: localVars[it] ?: JsonNull.INSTANCE
            }
            val fn = functions[fnName] ?: extractFunction(fnName).also {
                functions[fnName] = it
            }
            return fn(argValues)
        }

        throw InterpretException("Unsupported JS expression $expr")
    }

    @Throws(InterpretException::class)
    private fun extractObject(objName: String): JsonElement {
        val fields = code.find(OBJECT_RE % objName.escape(), "fields")
                ?: throw InterpretException("Failed to extract object $objName")

        val obj = JsonObject()
        for (match in fields.findAll(FIELDS_RE)) {
            val argNames = match.groupValue("args")!!.split(",")
            obj[match.groupValue("key")!!.removeQuotes()] = buildFunction(argNames, match.groupValue("code")!!)
        }
        return obj
    }

    @Throws(InterpretException::class)
    fun extractFunction(fnName: String): JsFunction {
        val escapedFnName = fnName.escape()
        val match = code.find(FN_RE % escapedFnName % escapedFnName % escapedFnName)
                ?: throw InterpretException("Couldn't find JS function $fnName")
        return buildFunction(match.groupValue("args")!!.split(","), match.groupValue("code")!!)
    }

    @Throws(InterpretException::class)
    fun callFunction(fnName: String, vararg args: Any): JsonElement = callFunction(fnName, args.toJsonArray())

    @Throws(InterpretException::class)
    private fun callFunction(fnName: String, args: JsonArray): JsonElement = extractFunction(fnName)(args)

    private fun buildFunction(argNames: List<String>, code: String): JsFunction = JsFunction { args ->
        val localVars = argNames.zip(args).toJsonObject()
        for (stmt in code.split(";")) {
            val (res, abort) = interpretStatement(stmt, localVars)
            if (abort) return@JsFunction res
        }
        JsonNull.INSTANCE
    }

    class InterpretException(override val message: String) : Exception(message)

    companion object {
        private val OPERATORS = linkedMapOf<String, OperatorFunction>(
                "|" to { l, r -> l or r },
                "^" to { l, r -> l xor r },
                "&" to { l, r -> l and r },
                ">>" to { l, r -> l shr r },
                "<<" to { l, r -> l shl r },
                "-" to { l, r -> l - r },
                "+" to { l, r -> l + r },
                "%" to { l, r -> l % r },
                "/" to { l, r -> l / r },
                "*" to { l, r -> l * r }
        ).toImmutableMap()
        private val ASSIGN_OPERATORS = OPERATORS.mapKeys { (key, _) -> "$key=" }.toMutableMap().apply {
            this["="] = { _, r -> r }
        }.toImmutableMap()

        @Language("RegExp")
        private const val NAME_RE = """[a-zA-Z_$][a-zA-Z_$0-9]*"""
        private const val ASSIGN_OPERATOR_RE = """(?x)(?<out>$NAME_RE)(?:\[(?<index>[^\]]+?)\])?\s*%s(?<expr>.*)$"""
        private const val VAR_RE = """^var\s"""
        private const val RETURN_RE = """^return(?:\s+|$)"""
        private const val VARIABLE_RE = """(?!if|return|true|false)(?<name>$NAME_RE)$"""
        private const val VARIABLE_INDEX_RE = """(?<in>$NAME_RE)\[(?<idx>.+)\]$"""
        private const val MEMBER_RE = """(?<var>$NAME_RE)(?:\.(?<member>[^(]+)|\[(?<member2>[^]]+)\])\s*(?:\(+(?<args>[^()]*)\))?$"""
        private const val FUNC_NAME_RE = """(?:[a-zA-Z$0-9]+|"[a-zA-Z$0-9]+"|'[a-zA-Z$0-9]+')"""
        private const val OBJECT_RE = """(?x)(?<!this\.)%s\s*=\s*\{\s*(?<fields>($FUNC_NAME_RE\s*:\s*function\s*\(.*?\)\s*\{.*?\}(?:,\s*)?)*)\}\s*;"""
        private const val FIELDS_RE = """(?x)(?<key>$FUNC_NAME_RE)\s*:\s*function\s*\((?<args>[a-z,]+)\)\{(?<code>[^}]+)\}"""
        private const val FN_CALL_RE = """^(?<func>$NAME_RE)\((?<args>[a-zA-Z0-9_$,]*)\)$"""
        private const val FN_RE = """(?x)(?:function\s+%s|[{;,]\s*%s\s*=\s*function|var\s+%s\s*=\s*function)\s*\((?<args>[^)]*)\)\s*\{(?<code>[^}]+)\}"""
    }
}